cmd/collect_signals/main.go

```diff
@@ -134,58 +218,20 @@ func main() {
 			"url": u.String(),
 		}).Debug("Parsed project url")
 
-		// TODO: Create a resolver in the project package.
-		r, err := ghRepoFactory.New(ctx, u)
-		if err != nil {
-			logger.WithFields(log.Fields{
-				"error": err,
-				"url":   u.String(),
-			}).Warning("Failed to create project")
-			continue
-			//os.Exit(1) // TODO: add a flag to continue or abort on failure
-			// TODO: we should have an error that indicates that the URL/Project
-			// should be skipped/ignored.
-		}
-
-		// TODO: p.URL() should be checked to see if it has already been processed.
-
-		// Collect the signals for the given project
-		logger.WithFields(log.Fields{
-			"url": u.String(),
-		}).Info("Collecting")
-		ss, err := collector.Collect(ctx, r)
-		if err != nil {
-			logger.WithFields(log.Fields{
-				"error": err,
-				"url":   r.URL().String(),
-			}).Error("Failed to collect signals for project")
-			os.Exit(1) // TODO: add a flag to continue or abort on failure
-		}
-
-		rec := out.Record()
-		for _, s := range ss {
-			if err := rec.WriteSignalSet(s); err != nil {
-				logger.WithFields(log.Fields{
-					"error": err,
-					"url":   r.URL().String(),
-				}).Error("Failed to write signal set")
-				os.Exit(1) // TODO: add a flag to continue or abort on failure
-			}
-		}
-		if err := rec.Done(); err != nil {
-			logger.WithFields(log.Fields{
-				"error":      err,
-				"repository": line,
-			}).Error("Failed to complete record")
-			os.Exit(1) // TODO: add a flag to continue or abort on failure
-		}
+		// Send the url to the workers
+		repos <- u
 	}
 	if err := scanner.Err(); err != nil {
 		logger.WithFields(log.Fields{
 			"error": err,
 		}).Error("Failed while reading input")
 		os.Exit(2)
 	}
+	// Close the repos channel to indicate that there is no more input.
+	close(repos)
```

Comment:
Should this be in a defer instead?

Reply:
No. It needs to happen before `wait()` defer would force it to happen afterwards.

At some point I will need to gracefully terminate other workers on a SIGTERM or a dead worker (unrecoverable error), but that will need more improvements.